designs/filewriter-refactor.md

```diff
@@ -0,0 +1,100 @@
+
+Finally, the `Write()` method writes files with unix permission bits 0777 before the umask is applied. This brings about risk as files are written with the execution bit for both group and other assuming a default umask of 022.
+
+## Proposal
+
+I propose we reduce the scope of the `FileWriter` to simply providing convinience methods for writing under a path context. Reducing the scope includes removal of the `CleanUp()` and `CleanUpTemp()` methods and any persistent or temporary file options. Additionally, we make clear consumers are responsible for cleaning any files they create with code docs on creation methods. If consumers wish to delete the whole directory they can use the standard libraries `os.RemoveAll(w.Dir())`. Additionally, any files created are assumed to be permanent. Consumers can explicitly define temporary file path contexts that are managed as independent contexts if they need to model temporary files. Finally, I propose a breakdown of interfaces to allow consumers to depend on only what they need and an introduction of a method set that compliments the standard libraries `io` package. Complienting the standard libraries `io` package ensures consumers can better satisfy the SRP as they can function without needing to manage file properties such as name and permissions.
+
+With respect to the default permissions assigned to files, I propose we change to using 0644 granting read and write to the owner and read permissions to group and other.
+
+## Implementation
+
+The following interfaces accept names as at least 1 parameter. The default implementation will prepend the path context to the name and return the `*os.File` type that can be used in `io` interfaces such as `io.Reader` and `io.Writer`.
+
+```go
+// Opener has the same behavior as os.Open().
+type Opener interface {
+    Open(name string) (*os.File, error)
+}
+
+// OpenFiler has the same behavior as os.OpenFile().
+type OpenFiler interface {
+    OpenFile(name string, flag int, perm os.FileMode) (*os.File, error)
+}
+
+// Creator has the same behavior as os.Create().
+type Creator interface {
+    Create(name string) (*os.File, error)
+}
+
+// WriteFiler performs the same function as the existing Write() call but mimics os.WriteFile() with the addition
+// of the path as a return value.
+type WriteFiler interface {
+    WriteFile(name string, data []byte, perm os.FileMode) (n int, path string, err error)
+}
+
+type FileWriter interface {
+    Opener
+    OpenFiler
+    Creator
+    WriteFiler
+
+    // Dir retrieves an absolute path to the current path.
+    Dir() string
+
+    // New creates a new FileWriter with subdir appended to the current path. Its the callers responsibility to delete
+    // the context if desired.
+    New(subdir string) FileWriter
```

Comment:
Are you suggesting using or mimicing? I'd be reluctant to use just because the `Open()` semantics on the `FS` interface return a file abstraction that can only be used for reading. 

Mimicing is probably a good idea though to be somewhat consistent. How about the following?

```
type FileWriter interface {
  Sub(subdir string) (FileWriter, error)
}
```